src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,688 @@
+    // @param algorithm the algorithm
+    // @param kdfParameters the algorithm parameters
+    private KDF(Delegate firstPairOfSpiAndProv, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        this.firstPairOfSpiAndProv = firstPairOfSpiAndProv;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        useFirstSpi();
+        return pairOfSpiAndProv.provider().getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     */
+    public KDFParameters getParameters() {
+        useFirstSpi();
+        return pairOfSpiAndProv.spi().engineGetParameters();
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     *
+     * @return a {@code KDF} object
+     *
+     * @implNote
+     * The JDK Reference Implementation additionally uses the
+     * {@code jdk.security.provider.preferred}
+     * {@link Security#getProperty(String) Security} property to determine
+     * the preferred provider order for the specified algorithm. This
+     * may be different than the order of providers returned by
+     * {@link Security#getProviders() Security.getProviders()}.
+     * See also the Cipher Transformations section of the {@extLink
+     * security_guide_jdk_providers JDK Providers} document for information
+     * on the transformation defaults used by JDK providers.
```

Comment:
This sentence is specific to the Cipher API, it does not apply here and should be removed.